accepted/00039-content-lifecycle-management.md

```diff
@@ -171,18 +175,44 @@ similar to the schedule definition of a repository sync.
 Adding a timer is optional.
 
 
-## Support restore
+## Restore functionality
+
+Support limited *Content* version tracking and restoring (or rolling back).
+
+The most important entity is the *Snapshot*. It has a version and it contains all the information needed for re-creating a *Build* in particular state in a future:
+* software channels reference (also the pointer which channel is the base (or "leader) in the *Build*
+* packages per channel reference
+* errata per channel
+* others (after we implement support for them): activation key, configuration channels, ...
+
+### Create Content Snapshot (backup)
+*Snapshot* is created after a successful *Project* *build*. It is populated with the references to channels and their packages/errata based on the *build*. Note that the software channels references point to the **source** channels of a project, not to the built ones.
+
+### Restore Content Snapshot
+Restoring a *Snapshot* to a *Environment*.
+
+The data from the *Snapshot* is evaluated and the *Content* is re-created from it (e.g. the software channel tree is re-created in given environment and populated by packages and errata from the *Snapshot*).
 
-As the current design does not allow a rollback of an *Environment* to previous versions of the *Content Build* the only chance to restore an *Environment* is to use the one from one stage above (restore).
+Technical note: For this, the adapted code for promoting *Environment*s could be used
 
-Example scenario:
-* 3 *Environment*s: `DEV`, `TEST`, `PRODUCTION`
-* Promoting *Content Build* from `DEV` to `TEST`
-* Testing the patches on systems in `TEST` fails, now we want to bring back the previous state:
+### Possible improvement: Using restore functionality for promoting
+The already existing *promote* functionality could be replaced by the *restore* code since it does the same.
 
-The solution can be:
- * 'restore' from `PRODUCTION` to `TEST`
+### Version management
+Keeping all versions of *Snapshot*s throughout the *Project* existence would be impractical for 2 reasons:
+* DB pollution (growing table for *Snapshot*)
+* integrity issues (it is likely that some entities referenced from very old *Snapshot*s will be deleted).
 
+The proposal is to keep only selected subset of the *Snapshot* versions per each *Environment*:
+* for the last *Environment*  in the *Path* (typically a "production" environment), keep 3 *Snapshot*s,
```

Comment:
Drawback of this nice differentiation of the *production* environment: what happen if a user already provides himself to create an additional environment at the end of the *Path* like an *archived* or *post-production* environment? I am afraid users would end up asking the possibility to *flag* somehow a certain environment as a *production* environment and keep 3 *Snapshot*s for it instead. I am more for keeping 1 for each environment, and adding a documentation best-practice of creating always an additional environment at the end of the *Path*. Combining the two I see more flexibility and we don't need to distinguish any environment in anyway.

Reply:
I added an alternative proposal with your suggestion!